readthedocs/storage/security_token_service.py

```diff
@@ -0,0 +1,154 @@
+"""
+Module to interact with AWS STS (Security Token Service) to assume a role and get temporary scoped credentials.
+
+This is mainly used to generate temporary credentials to interact with S3 buckets from the builders.
+
+In order to make use of STS, we need:
+
+- Create a role in IAM with a trusted entity type set to the AWS account that is going to be used to generate the temporary credentials.
+- A policy that allows access to all S3 buckets and paths that are going to be used.
+  Which should be attached to the role.
+- The permissions of the temporary credentials are the result of the intersection of the role policy and the inline policy that is passed to the AssumeRole API.
+  This means that the inline policy can be used to limit the permissions of the temporary credentials, but not to expand them.
+
+See:
+
+- https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html
+- https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_sts-comparison.html
+- https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_assumerole.html
+"""
+
+import json
+from dataclasses import dataclass
+
+import boto3
+import structlog
+from django.conf import settings
+
+
+log = structlog.get_logger(__name__)
+
+
+class AWSTemporaryCredentialsError(Exception):
+    """Exception raised when there is an error getting AWS S3 credentials."""
+
+
+@dataclass
+class AWSTemporaryCredentials:
+    """Dataclass to hold AWS temporary credentials."""
+
+    access_key_id: str
+    secret_access_key: str
+    session_token: str | None
+
+
+def get_sts_client():
+    return boto3.client(
+        "sts",
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+        # TODO: should this be its own setting?
+        region_name=settings.AWS_S3_REGION_NAME,
+    )
+
+
+def get_s3_scoped_credentials(
+    *, project, version, session_id=None, duration=60 * 15
+) -> AWSTemporaryCredentials:
+    """
+    :param project: The project to get the credentials for.
+    :param version: The version to get the credentials for.
+    :param session_id: A unique identifier to add to the name of the role session.
+     The name of the session always includes the project and version slug (rtd-{project}-{version}),
+     if session_id is given, the name of the session would be "rtd-{session_id}-{project}-{version}".
+     AWS limits the session name to 64 characters, so if the session_id is too long, it will be truncated.
+     For example, for a token used in a build, a good session_id is the ID of the build.
+    :duration: The duration of the credentials in seconds. Default is 15 minutes.
+     Note that the minimum duration time is 15 minutes and the maximum is given by the role (defaults to 1 hour).
```

Comment:
We likely need more than this for some builds -- I'd say up to 3 hours just to be safe. We really don't want long builds failing because of this all the way at the end.